File: doma-core/src/main/java/org/seasar/doma/jdbc/criteria/query/SelectBuilder.java
package org.seasar.doma.jdbc.criteria.query;

import static java.util.stream.Collectors.toList;

import java.util.List;
import java.util.Objects;
import java.util.function.Function;
import org.seasar.doma.internal.jdbc.sql.PreparedSqlBuilder;
import org.seasar.doma.internal.util.Pair;
import org.seasar.doma.jdbc.Config;
import org.seasar.doma.jdbc.PreparedSql;
import org.seasar.doma.jdbc.SqlKind;
import org.seasar.doma.jdbc.SqlLogType;
import org.seasar.doma.jdbc.criteria.context.Criterion;
import org.seasar.doma.jdbc.criteria.context.Join;
import org.seasar.doma.jdbc.criteria.context.JoinKind;
import org.seasar.doma.jdbc.criteria.context.OrderByItem;
import org.seasar.doma.jdbc.criteria.context.SelectContext;
import org.seasar.doma.jdbc.criteria.context.SetOperationContext;
import org.seasar.doma.jdbc.criteria.expression.AggregateFunction;
import org.seasar.doma.jdbc.criteria.metamodel.EntityMetamodel;
import org.seasar.doma.jdbc.criteria.metamodel.PropertyMetamodel;
import org.seasar.doma.jdbc.criteria.option.DistinctOption;
import org.seasar.doma.jdbc.criteria.option.ForUpdateOption;

public class SelectBuilder {
  private final SelectContext context;
  private final Function<String, String> commenter;
  private final PreparedSqlBuilder buf;
  private final AliasManager aliasManager;
  private final BuilderSupport support;
  private final CriteriaBuilder criteriaBuilder;

  public SelectBuilder(
      Config config,
      SelectContext context,
      Function<String, String> commenter,
      SqlLogType sqlLogType) {
    this(
        config,
        context,
        commenter,
        new PreparedSqlBuilder(config, SqlKind.SELECT, sqlLogType),
        AliasManager.create(config, context));
  }

  public SelectBuilder(
      Config config,
      SelectContext context,
      Function<String, String> commenter,
      PreparedSqlBuilder buf,
      AliasManager aliasManager) {
    Objects.requireNonNull(config);
    Objects.requireNonNull(config);
    this.context = Objects.requireNonNull(context);
    this.commenter = Objects.requireNonNull(commenter);
    this.buf = Objects.requireNonNull(buf);
    this.aliasManager = Objects.requireNonNull(aliasManager);
    support = new BuilderSupport(config, commenter, buf, aliasManager);
    criteriaBuilder = config.getDialect().getCriteriaBuilder();
  }

  public PreparedSql build() {
    interpret();
    return buf.build(commenter);
  }

  void interpret() {
    with();
    select();
    from();
    join();
    where();
    groupBy();
    having();
    orderBy();
    offsetAndFetch();
    forUpdate();
  }

  private void with() {
    support.with(context.withContexts);
  }

  private void select() {
    buf.appendSql("select ");

    if (context.distinct == DistinctOption.Kind.BASIC) {
      buf.appendSql("distinct ");
    }

    List<PropertyMetamodel<?>> propertyMetamodels = context.getProjectionPropertyMetamodels();
    if (propertyMetamodels.isEmpty()) {
      buf.appendSql("*");
    } else {
      for (PropertyMetamodel<?> propertyMetamodel : propertyMetamodels) {
        selectColumn(propertyMetamodel);
        buf.appendSql(", ");
      }
      buf.cutBackSql(2);
    }
  }

  private void from() {
    buf.appendSql(" from ");
    SetOperationContext<?> setOperationContext =
        context.setOperationContextForSubQuery.orElse(null);
    if (setOperationContext != null) {
      subQuery(context.entityMetamodel, setOperationContext, aliasManager);
    } else {
      table(context.entityMetamodel);
    }
    if (context.forUpdate != null) {
      ForUpdateOption option = context.forUpdate.option;
      criteriaBuilder.lockWithTableHint(buf, option, this::column);
    }
  }

  private void join() {
    if (!context.joins.isEmpty()) {
      for (Join join : context.joins) {
        if (join.kind == JoinKind.INNER) {
          buf.appendSql(" inner join ");
        } else if (join.kind == JoinKind.LEFT) {
          buf.appendSql(" left outer join ");
        }
        table(join.entityMetamodel);
        if (!join.on.isEmpty()) {
          buf.appendSql(" on (");
          int index = 0;
          for (Criterion criterion : join.on) {
            visitCriterion(index, criterion);
            index++;
            buf.appendSql(" and ");
          }
          buf.cutBackSql(5);
          buf.appendSql(")");
        }
      }
    }
  }

  private void where() {
    if (!context.where.isEmpty()) {
      buf.appendSql(" where ");
      int index = 0;
      for (Criterion criterion : context.where) {
        visitCriterion(index++, criterion);
        buf.appendSql(" and ");
      }
      buf.cutBackSql(5);
    }
  }

  private void groupBy() {
    if (context.groupBy.isEmpty()) {
      List<PropertyMetamodel<?>> propertyMetamodels = context.getProjectionPropertyMetamodels();
      if (propertyMetamodels.stream().anyMatch(p -> p instanceof AggregateFunction<?>)) {
        List<PropertyMetamodel<?>> groupKeys =
            propertyMetamodels.stream()
                .filter(p -> !(p instanceof AggregateFunction<?>))
                .collect(toList());
        context.groupBy.addAll(groupKeys);
      }
    }
    if (!context.groupBy.isEmpty()) {
      buf.appendSql(" group by ");
      for (PropertyMetamodel<?> p : context.groupBy) {
        column(p);
        buf.appendSql(", ");
      }
      buf.cutBackSql(2);
    }
  }

  private void having() {
    if (!context.having.isEmpty()) {
      buf.appendSql(" having ");
      int index = 0;
      for (Criterion criterion : context.having) {
        visitCriterion(index++, criterion);
        buf.appendSql(" and ");
      }
      buf.cutBackSql(5);
    }
  }

  private void orderBy() {
    if (!context.orderBy.isEmpty()) {
      buf.appendSql(" order by ");
      for (Pair<OrderByItem, String> pair : context.orderBy) {
        pair.fst.accept(
            new OrderByItem.Visitor() {

              @Override
              public void visit(OrderByItem.Name name) {
                column(name.value);
              }

              @Override
              public void visit(OrderByItem.Index index) {
                buf.appendSql(String.valueOf(index.value));
              }
            });
        buf.appendSql(" " + pair.snd + ", ");
      }
      buf.cutBackSql(2);
    }
  }

  private void offsetAndFetch() {
    if (context.offset == null && context.limit == null) {
      return;
    }
    int offset = (context.offset == null || context.offset < 0) ? 0 : context.offset;
    int limit = (context.limit == null || context.limit < 0) ? 0 : context.limit;
    criteriaBuilder.offsetAndFetch(buf, offset, limit);
  }

  private void forUpdate() {
    if (context.forUpdate != null) {
      ForUpdateOption option = context.forUpdate.option;
      criteriaBuilder.forUpdate(buf, option, this::column, aliasManager);
    }
  }

  private void table(EntityMetamodel<?> entityMetamodel) {
    support.table(entityMetamodel);
  }

  private void subQuery(
      EntityMetamodel<?> entityMetamodel,
      SetOperationContext<?> setOperationContext,
      AliasManager aliasManager) {
    support.subQuery(entityMetamodel, setOperationContext, aliasManager);
  }

  private void column(PropertyMetamodel<?> propertyMetamodel) {
    support.column(propertyMetamodel);
  }

  private void selectColumn(PropertyMetamodel<?> propertyMetamodel) {
    support.selectColumn(propertyMetamodel);
  }

  private void visitCriterion(int index, Criterion criterion) {
    support.visitCriterion(index, criterion);
  }
}

File: doma-criteria/src/main/kotlin/org/seasar/doma/criteria/query/SelectBuilder.kt
package org.seasar.doma.criteria.query

import org.seasar.doma.criteria.context.Criterion
import org.seasar.doma.criteria.context.JoinKind
import org.seasar.doma.criteria.context.Projection
import org.seasar.doma.criteria.context.SelectContext
import org.seasar.doma.internal.jdbc.sql.PreparedSqlBuilder
import org.seasar.doma.jdbc.PreparedSql
import org.seasar.doma.jdbc.SqlKind
import org.seasar.doma.jdbc.SqlLogType
import org.seasar.doma.jdbc.entity.EntityPropertyType
import org.seasar.doma.jdbc.entity.EntityType

class SelectBuilder(
    private val context: SelectContext,
    private val commenter: (String) -> String,
    private val buf: PreparedSqlBuilder,
    aliasManager: AliasManager
) {

    constructor(context: SelectContext, commenter: (String) -> String, logType: SqlLogType) :
            this(context, commenter, PreparedSqlBuilder(context.config, SqlKind.SELECT, logType), AliasManager(context))

    private val config = context.config
    private val support = BuilderSupport(config, commenter, buf, aliasManager)
    fun build(): PreparedSql {
        interpretContext()
        return buf.build(commenter)
    }

    internal fun interpretContext() {
        buf.appendSql("select ")
        if (context.distinct) {
            buf.appendSql("distinct ")
        }
        when (val projection = context.projection) {
            is Projection.All -> {
                context.getProjectionTargets().forEach {
                    it.entityPropertyTypes.forEach { prop ->
                        column(prop)
                        buf.appendSql(", ")
                    }
                }
                buf.cutBackSql(2)
            }
            is Projection.Asterisk -> {
                buf.appendSql("*")
            }
            is Projection.Single -> {
                column(projection.propType)
            }
            is Projection.Pair -> {
                column(projection.first)
                buf.appendSql(", ")
                column(projection.second)
            }
            is Projection.List -> {
                projection.propTypes.forEach {
                    column(it)
                    buf.appendSql(", ")
                }
                buf.cutBackSql(2)
            }
        }
        buf.appendSql(" from ")
        table(context.entityType)
        if (context.joins.isNotEmpty()) {
            context.joins.forEach { join ->
                when (join.kind) {
                    JoinKind.INNER -> buf.appendSql(" inner join ")
                    JoinKind.LEFT -> buf.appendSql(" left outer join ")
                }
                table(join.entityType)
                if (join.on.isNotEmpty()) {
                    buf.appendSql(" on (")
                    join.on.forEachIndexed { index, c ->
                        visitCriterion(index, c)
                        buf.appendSql(" and ")
                    }
                    buf.cutBackSql(5)
                    buf.appendSql(")")
                }
            }
        }
        if (context.where.isNotEmpty()) {
            buf.appendSql(" where ")
            context.where.forEachIndexed { index, c ->
                visitCriterion(index, c)
                buf.appendSql(" and ")
            }
            buf.cutBackSql(5)
        }
        if (context.groupBy.isNotEmpty()) {
            buf.appendSql(" group by ")
            context.groupBy.forEach { p ->
                column(p)
                buf.appendSql(", ")
            }
            buf.cutBackSql(2)
        }
        if (context.having.isNotEmpty()) {
            buf.appendSql(" having ")
            context.having.forEachIndexed { index, c ->
                visitCriterion(index, c)
                buf.appendSql(" and ")
            }
            buf.cutBackSql(5)
        }
        if (context.orderBy.isNotEmpty()) {
            buf.appendSql(" order by ")
            context.orderBy.forEach { (p, sort) ->
                column(p)
                buf.appendSql(" $sort")
                buf.appendSql(", ")
            }
            buf.cutBackSql(2)
        }
        context.limit?.let {
            buf.appendSql(" limit $it")
        }
        context.offset?.let {
            buf.appendSql(" offset $it")
        }
        context.forUpdate?.let {
            buf.appendSql(" for update")
            if (it.nowait) {
                buf.appendSql(" nowait")
            }
        }
    }

    private fun table(entityType: EntityType<*>) {
        support.table(entityType)
    }

    private fun column(propType: EntityPropertyType<*, *>) {
        support.column(propType)
    }

    private fun visitCriterion(index: Int, c: Criterion) {
        support.visitCriterion(index, c)
    }
}
